workspaces/optic-engine/src/events/http_interaction.rs

```diff
@@ -32,7 +32,7 @@ pub struct Request {
   pub path: String,
   pub headers: ArbitraryData,
   // #[serde(skip)]
-  pub query: ArbitraryData,
+  pub query: QueryParametersData,
```

Comment:
What's the intent of this change? In the original design, the query string should be saved raw in asText, and either at capture time or pre-diffing should be parsed by a pluggable parser into a JSON object. Currently we have 2 capture sources; the local cli proxy and the rust proxy, with more coming. I would probably not use the existing query string parser implementation the proxy is using, and instead use the node url module by default but I would leave a pathway for it to be changed. The contract is that the query string should parse into a diffable object, so it can use the existing diffing logic and commands (aside from the new command to replace the root shape ID)

Reply:
It's a simple wrapper around `ArbitraryData` (flattened for serde, so invisible to the outside), so we can implement a separate way of parsing the `asText` of a query parameter to a `BodyDescriptor`. So why even bother with that, if the query string is also available as shape hash, json?

During the exploration and forming of the bet, we ran into the giant tyre fire that is the world of parsing query params, with sometimes not even respecting the RFC's for safe urls, using different key-value delimiters, etc. It's obvious that at some point, the user should be able to provide their own parsing strategy implementation, while the Optic one should be the most base one that makes sure all params make it into the documentation, but all as flat strings.

For the current bet, we're strictly on taking that most base way of parsing query parameters, with no other kind of structure supported. Not only to limit the complexity on the domain side, but also all the UI and UX work needed to establish base support for query parameters.

Now this was only a drat, but thinking ahead of how we would eventually move to a world where query parameters are parsed outside from the domain (like what is currently happening in the capture proxy with the `DefaultQueryParser`), I reckoned it might make sense to:

- For now only read query parameters from `asText` and perform the most base way of parsing query parameters (building a `Vec<(String, String)>` to represent all the entries) on the Rust side. That way, we know no matter the source of the capture, our base way of parsing query parameters is the same and known to the Engine. We'll know we won't get any more complicated shapes, we'll know what to expect building our first UI, etc.
- Later, start attempt to read query parameters from the ShapeHash or Json first, and treating that as custom parsed query strings (even if those parsers are default Optic ones). By that point we should expect pretty much any shape, with the additional UI and UX work on that to follow as well.

Writing this down now, I do wonder if that simple mechanism for choosing default parser vs. custom parser will work for non-local captures, where only the shape hash might be the acceptable way of transmitting that data. Perhaps not. If that's the case, we might have to come up with something more elaborate. Regardless, I reckon it's a good idea for the Rust codebase to be the place where the original query string -> parameters parser lives, just to make sure it's consistent everywhere (Node qs parsers have weird idiosyncrasies as well, as do the ones in pretty much any other language, and probably the Rust one too, but the more compelling the idea to make sure to use just one implementation).